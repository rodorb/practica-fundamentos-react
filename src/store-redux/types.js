export const ADVERTISEMENTS_REQUEST = 'ADVERTISEMENTS_REQUEST';
export const ADVERTISEMENTS_SUCCESS = 'ADVERTISEMENTS_SUCCESS';
export const ADVERTISEMENTS_FAILURE = 'ADVERTISEMENTS_FAILURE';

export const AD_REQUEST = 'AD_REQUEST';
export const AD_SUCCESS = 'AD_SUCCESS';
export const AD_FAILURE = 'AD_FAILURE';

export const AD_CREATION_REQUEST = 'AD_CREATION_REQUEST';
export const AD_CREATION_SUCCESS = 'AD_CREATION_SUCCESS';
export const AD_CREATION_FAILURE = 'AD_CREATION_FAILURE';

export const AD_DELETION_REQUEST = 'AD_DELETION_REQUEST';
export const AD_DELETION_SUCCESS = 'AD_DELETION_SUCCESS';
export const AD_DELETION_FAILURE = 'AD_DELETION_FAILURE';


export const TAGS_REQUEST = 'TAGS_REQUEST';
export const TAGS_SUCCESS = 'TAGS_SUCCESS';
export const TAGS_FAILURE = 'TAGS_FAILURE';

export const AUTHENTICATION_LOGIN_REQUEST = 'AUTHENTICATION_LOGIN_REQUEST';
export const AUTHENTICATION_LOGIN_SUCCESS = 'AUTHENTICATION_LOGIN_SUCCESS';
export const AUTHENTICATION_LOGIN_FAILURE = 'AUTHENTICATION_LOGIN_FAILURE';

export const AUTHENTICATION_LOGOUT_SUCCESS = 'AUTHENTICATION_LOGOUT_SUCCESS';

export const UI_RESET_ERROR = 'UI_RESET_ERROR';