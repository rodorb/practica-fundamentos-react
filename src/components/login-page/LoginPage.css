.loginPage {
    padding: 3rem;
    margin: 5rem;
    align-self: center;
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
}

.loginPage-title {
    align-items: center;
    display: flex;
    height: 64px;
    justify-content: center;
    margin-bottom: 10px;
    color: white;
}


/* 


.loginPage-error {
    background-color: rgb(224, 36, 94, 0.1);
    border: 1px solid rgba(224, 36, 94, 0.2);
    border-radius: 5px;
    color: rgb(224, 36, 94);
    margin-top: 20px;
    padding: 10px 15px;
}

.loginForm-field {
    padding: 10px 0;
}

.loginForm-submit {
    margin-top: 10px;
    width: 100%;
} */