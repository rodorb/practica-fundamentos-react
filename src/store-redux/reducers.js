import {
    ADVERTISEMENTS_FAILURE,
    ADVERTISEMENTS_REQUEST,
    ADVERTISEMENTS_SUCCESS,
    AD_CREATION_FAILURE,
    AD_CREATION_REQUEST,
    AD_CREATION_SUCCESS,
    AD_DELETION_FAILURE,
    AD_DELETION_REQUEST,
    AD_DELETION_SUCCESS,
    AD_FAILURE,
    AD_REQUEST,
    AD_SUCCESS,
    AUTHENTICATION_LOGIN_FAILURE,
    AUTHENTICATION_LOGIN_REQUEST,
    AUTHENTICATION_LOGIN_SUCCESS,
    AUTHENTICATION_LOGOUT_SUCCESS,
    TAGS_FAILURE,
    TAGS_REQUEST,
    TAGS_SUCCESS,
    UI_RESET_ERROR
} from "./types";


export const defaultState = {

    authentication: false,
    advertisements: {
        data: [],
        loaded: false
    },
    tags: {
        data: [],
        loaded: false
    },
    ui: {
        isLoading: false,
        error: null
    }
};

export const authentication = (state = defaultState.authentication, action) => {
    switch (action) {
        case AUTHENTICATION_LOGIN_SUCCESS:
            return true;
        case AUTHENTICATION_LOGOUT_SUCCESS:
            return false;
        default:
            return state;
    }

}

export const advertisements = (state = defaultState.advertisements, action) => {
    switch (action) {
        case ADVERTISEMENTS_SUCCESS:
            return { loaded: true, data: action.payload };
        case AD_SUCCESS:
            return { loaded: true, data: [...state.data, action.payload] };
        case AD_CREATION_SUCCESS:
            return { loaded: true, data: [action.payload, ...state.data] }; //para que el anuncio creado estÃ© en primera posicion en el listado
        case AD_DELETION_SUCCESS:
            return { loaded: true, data: state.data.filter((ad) => { return ad.id !== action.payload }) }; //en el payload devolvere el id del anuncio borrado
        default:
            return state;
    }
}

export const tags = (state = defaultState.tags, action) => {
    switch (action) {
        case TAGS_SUCCESS:
            return { loaded: true, data: action.payload };
        default:
            return state;
    }
}

export const ui = (state = defaultState.ui, action) => {
    switch (action) {
        case AUTHENTICATION_LOGIN_REQUEST:
        case ADVERTISEMENTS_REQUEST:
        case AD_CREATION_REQUEST:
        case AD_DELETION_REQUEST:
        case TAGS_REQUEST:
        case AD_REQUEST:
            return {...state, isLoading: true, error: null };
        case AUTHENTICATION_LOGIN_SUCCESS:
        case ADVERTISEMENTS_SUCCESS:
        case AD_CREATION_SUCCESS:
        case AD_DELETION_SUCCESS:
        case TAGS_SUCCESS:
        case AD_SUCCESS:
            return {...state, isLoading: false };
        case AUTHENTICATION_LOGIN_FAILURE:
        case ADVERTISEMENTS_FAILURE:
        case AD_CREATION_FAILURE:
        case AD_DELETION_FAILURE:
        case TAGS_FAILURE:
        case AD_FAILURE:
            return {...state, isLoading: false, error: action.payload };
        case UI_RESET_ERROR:
            return {...state, error: null }
        default:
            return state;
    }
}