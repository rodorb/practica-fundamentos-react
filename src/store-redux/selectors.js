export const getIsLogged = (state) => { return state.authentication };
export const getAdvertisements = (state) => { return state.advertisements.data };
export const getAreAdvertisementsLoaded = (state) => { return state.advertisements.loaded };
export const getAdvert = (advertId) => {
    return (state) => {
        return state.advertisements.data.find((ad) => ad.id === advertId);
    }
};
export const getTags = (state) => { return state.tags.data };
export const getAreTagsLoaded = (state) => { return state.tags.loaded };

export const getUi = (state) => { return state.ui };