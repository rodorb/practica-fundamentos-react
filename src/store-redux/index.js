import { applyMiddleware, combineReducers, createStore } from 'redux';
import { composeWithDevTools } from '@redux-devtools/extension';
import thunk from 'redux-thunk';
import * as reducers from './reducers';
import AdvertsService from '../components/adverts/service/AdvertsService';
import AuthService from '../components/auth/AuthService';
import { AD_CREATION_SUCCESS, AD_DELETION_SUCCESS, AUTHENTICATION_LOGIN_SUCCESS, AUTHENTICATION_LOGOUT_SUCCESS } from './types';
const api = { advertsService: AdvertsService, authService: AuthService };

const failureRedirections =
    (history, redirections) => _store => next => action => {
        const result = next(action);
        if (action.error) {
            const redirection = redirections[action.payload.status];
            if (redirection) {
                history.push(redirection);
            }
        }

        return result;
    };

const successRedirections =
    (history, redirections) => _store => next => action => {
        const result = next(action);
        const redirection = redirections[action.type];
        if (redirection) {
            redirection(history, action.payload);
        }

        return result;
    };

const storeSetup = (preloadedState, { history }) => {
    const middlewares = [
        thunk.withExtraArgument({ api, history }),
        failureRedirections(history, { 401: '/login', 404: '/404' }),
        successRedirections(history, {
            [AUTHENTICATION_LOGIN_SUCCESS]: history => {
                const redirectTo = history?.location?.state?.redirectTo?.pathname || '/';
                history.replace(redirectTo);
            },
            [AUTHENTICATION_LOGOUT_SUCCESS]: history => history.replace('/login'),
            [AD_CREATION_SUCCESS]: (history, payload) => history.push(`/adverts/${payload.id}`),
            [AD_DELETION_SUCCESS]: (history) => history.push(`/`)
        }),
    ];
    return createStore(
        combineReducers(reducers),
        preloadedState,
        composeWithDevTools(applyMiddleware(...middlewares)))
};

export default storeSetup;