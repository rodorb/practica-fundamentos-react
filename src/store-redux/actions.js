import { getAdvert, getAreAdvertisementsLoaded, getAreTagsLoaded } from "./selectors";
import { ADVERTISEMENTS_FAILURE, ADVERTISEMENTS_REQUEST, ADVERTISEMENTS_SUCCESS, AD_CREATION_REQUEST, AD_CREATION_SUCCESS, AD_DELETION_REQUEST, AD_DELETION_SUCCESS, AD_REQUEST, AUTHENTICATION_LOGIN_FAILURE, AUTHENTICATION_LOGIN_REQUEST, AUTHENTICATION_LOGIN_SUCCESS, AUTHENTICATION_LOGOUT_SUCCESS, TAGS_FAILURE, TAGS_REQUEST, TAGS_SUCCESS, UI_RESET_ERROR } from "./types"

///LOGIN
export const loginRequest = () => {
    return {
        type: AUTHENTICATION_LOGIN_REQUEST
    }
};
export const loginSucces = () => {
    return {
        type: AUTHENTICATION_LOGIN_SUCCESS
    }
};
export const loginFailure = (error) => {
    return {
        type: AUTHENTICATION_LOGIN_FAILURE,
        payload: error,
        error: true
    }
};
export const logoutSuccess = () => ({
    type: AUTHENTICATION_LOGOUT_SUCCESS,
});

export const login = (credentials) => {
    return async(dispatch, _getState, { api }) => {
        const { authService } = api;
        dispatch(loginRequest());
        try {
            await authService.login(credentials);
            dispatch(loginSucces());
            //TODO: hacer la redireccion
        } catch (error) {
            dispatch(loginFailure(error));
        }
    }
};

export const logout = () => {
    return function(dispatch, _getState, { api }) {
        const { authService } = api;
        return authService.logout().then(() => {
            dispatch(logoutSuccess());
        });
    };
};

//ALL ADVERTISEMENTS
export const advertsRequest = () => {
    return {
        type: ADVERTISEMENTS_REQUEST
    }
};

export const advertsSuccess = (adverts) => {
    return {
        type: ADVERTISEMENTS_SUCCESS,
        payload: adverts
    }
};

export const advertsFailure = (error) => {
    return {
        type: ADVERTISEMENTS_FAILURE,
        payload: error,
        error: true
    }
};


export const advertisements = () => {
    return async(dispatch, _getState, { api }) => {
        const { advertsService } = api;
        //Lógica para que al volver a la pagina del listado (siempre y cuando no recargue la pagina)
        //no se vuelva a hacer la peticion al API, si no que coja los datos del estado de redux
        const advertismentsLoaded = getAreAdvertisementsLoaded(_getState());
        if (advertismentsLoaded) return;
        try {
            dispatch(advertsRequest());
            const advertisements = await advertsService.getAllAdvertisements();
            dispatch(advertsSuccess(advertisements));
        } catch (error) {
            dispatch(advertsFailure(error));
        }
    }
};



//GET ADVERT
export const advertRequest = () => {
    return {
        type: AD_REQUEST
    }
};

export const advertSuccess = (advert) => {
    return {
        type: AD_REQUEST,
        payload: advert
    }
};

export const advertFailure = (error) => {
    return {
        type: AD_REQUEST,
        payload: error,
        error: true
    }
};

export const advertisement = (advertId) => {
    return (dispatch, _getState, { api }) => {
        const { advertsService } = api;
        //Lógica apra que solo haga la llamad al API cuando se recargue la pagina,
        //no cuando vaya del lsitado de anuncios al detalle de unanuncio que ya se encuentra
        //en el store
        const advertLoaded = getAdvert(advertId)(_getState());
        if (advertLoaded) return;
        try {
            dispatch(advertRequest());
            const advert = advertsService.getAdvert(advertId);
            dispatch(advertSuccess(advert))

        } catch (error) {
            dispatch(advertFailure(error));
        }
    }
};

//ADVERT CREATION

export const advertCreationRequest = () => {
    return {
        type: AD_CREATION_REQUEST
    };
};

export const advertCreationSuccess = (advert) => {
    return {
        type: AD_CREATION_SUCCESS,
        payload: advert
    };
};

export const advertCreationFailure = (error) => {
    return {
        type: AD_CREATION_REQUEST,
        payload: error,
        error: true
    };
};


export const advertCreation = (body) => {
    return async(dispatch, _getState, { api }) => {
        const { advertsService } = api;
        try {
            dispatch(advertCreationRequest());
            const { id } = await advertsService.createAdvert(body);
            const createdAdvert = await advertsService.getAdvert(id);
            dispatch(advertCreationSuccess(createdAdvert));
        } catch (error) {
            dispatch(advertCreationFailure(error));
        }
    }
};

//ADVERT DELETION

export const advertDeletionRequest = () => {
    return {
        type: AD_DELETION_REQUEST
    };
};

export const advertDeletionSuccess = (advertId) => {
    return {
        type: AD_DELETION_SUCCESS,
        payload: advertId
    };
};

export const advertDeletionFailure = (error) => {
    return {
        type: AD_DELETION_REQUEST,
        payload: error,
        error: true
    };
};


export const advertDeletion = (advertId) => {
    return async(dispatch, _getState, { api }) => {
        const { advertsService } = api;
        try {
            dispatch(advertDeletionRequest());
            await advertsService.deleteAdvert(advertId);
            dispatch(advertDeletionSuccess(advertId));
        } catch (error) {
            dispatch(advertDeletionFailure(error));
        }
    }
};


//TAGS
export const tagsRequest = () => {
    return {
        type: TAGS_REQUEST
    };
};

export const tagsSuccess = (tags) => {
    return {
        type: TAGS_SUCCESS,
        payload: tags
    };
};

export const tagsFailure = (error) => {
    return {
        type: TAGS_FAILURE,
        payload: error,
        error: true
    };
};


export const allTags = () => {
    return async(dispatch, _getState, { api }) => {
        const { advertsService } = api;
        const tagsLoaded = getAreTagsLoaded(_getState());
        if (tagsLoaded) return;
        try {
            dispatch(tagsRequest());
            const tags = await advertsService.getTags();
            dispatch(tagsSuccess(tags));
        } catch (error) {
            dispatch(tagsFailure(error));
        }
    }
};


//UI
export const uiResetError = () => ({
    type: UI_RESET_ERROR,
});