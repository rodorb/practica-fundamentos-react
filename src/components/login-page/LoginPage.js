import { useState } from "react";
import { FormField } from "../../shared/components/ui-components/FormField";
import { Page } from "../layout/Page";
import './LoginPage.css'
import ButtonBootstrap from 'react-bootstrap/Button';
import Spinner from 'react-bootstrap/Spinner';
import Toast from 'react-bootstrap/Toast';
import { login } from "../../store-redux/actions";
import { useDispatch, useSelector } from "react-redux";
import { getUi } from "../../store-redux/selectors";



export const LoginPage = () => {
    const dispatch = useDispatch();
    const { error, isLoading } = useSelector(getUi);
   
    const [userCredentials, setUserCredentials] = useState({
        email: '',
        password:  '',
        rememberCredentials:  false
    });

    const { email, password, rememberCredentials } = userCredentials;

    const handleInputChange = ({ target: { value, name, type, checked } }) => {
        // const evtTarget = event?.target;
        setUserCredentials(credentials => ({
            ...credentials,
            [name]: type === 'checkbox' ?
            checked : value
        }))
    };


    const handleFormSubmit = async (event) => {
        event.preventDefault();
        dispatch(login(userCredentials));
        

    }

    const isButtonDisabled = !email || !password || isLoading;

    return (
      <Page>
        <div className="loginPage">
          <h1 className="loginPage-title">Log in to Nodepop</h1>
          <form className="loginForm" onSubmit={handleFormSubmit}>
            <FormField
              type="text"
              name="email"
              label="User Email"
              className="loginForm-field"
              value={email}
              onChange={handleInputChange}
              // ref={ref}
            />
            <FormField
              type="password"
              name="password"
              label="Password"
              className="loginForm-field"
              value={password}
              onChange={handleInputChange}
            />

            <input
              type="checkbox"
              name="rememberCredentials"
              checked={rememberCredentials}
              value="rememberCredentials"
              onChange={handleInputChange}
            />
            <label htmlFor="rememberCredentials">Recordar credenciales</label>
           


            <div className="formButton">
                <ButtonBootstrap size="lg" className="submit-form-button" variant="primary" type="submit" disabled={isButtonDisabled}>
                  Acceder                   
                </ButtonBootstrap>
              </div>
          </form>

          {isLoading && (
            <div>
             <Spinner animation="border" role="status" variant="warning" >
              <span className="visually-hidden">Loading...</span>
            </Spinner>
            </div>
          )}

          {error && (
            <Toast bg="danger">
            <Toast.Header>
              <strong className="me-auto">Error</strong>
            </Toast.Header>
            <Toast.Body>Se ha producido un error en la aplicaci√≥n.</Toast.Body>
          </Toast>
          )}

        </div>
      </Page> 
        
      );

}