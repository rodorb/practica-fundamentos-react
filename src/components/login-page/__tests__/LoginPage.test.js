import { fireEvent, render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import {LoginPage} from '../LoginPage';
import { login, uiResetError } from '../../../store-redux/actions';

jest.mock('../../../store-redux/actions');

describe('LoginPage', () => {
  test('snapshot', () => {
    const state = {
      ui: {
        error: null,
        isLoading: false,
      },
    };
    const store = {
      getState: () => state,
      dispatch: jest.fn(),
      subscribe: () => {},
    };
    const { container } = render(
      <Provider store={store}>
        <LoginPage />
      </Provider>,
    );
    expect(container).toMatchSnapshot();
  });

  test('should authLogin action', () => {
    login.mockReturnValue('action');
    const state = {
      ui: {
        error: null,
        isLoading: false,
      },
    }; 
    const username = 'asd';
    const password = '1234';
    const store = {
      getState: () => state,
      dispatch: jest.fn(),
      subscribe: () => {},
    };
    render(
      <Provider store={store}>
        <LoginPage />
      </Provider>,
    );
    const usernameInput = screen.getByLabelText(/User Email/);
    const passwordInput = screen.getByLabelText(/Password/); 
    const submitButton = screen.getByRole('button');

    expect(submitButton).toBeDisabled();
    fireEvent.change(usernameInput, { target: { value: username } });
    fireEvent.change(passwordInput, { target: { value: password } });
    expect(submitButton).not.toBeDisabled();

    fireEvent.click(submitButton);
    // const credentials = login.mock.calls[0][0];
    // expect(credentials).toMatchObject({ username, password });
    expect(store.dispatch).toHaveBeenCalledWith('action');
  });

});
