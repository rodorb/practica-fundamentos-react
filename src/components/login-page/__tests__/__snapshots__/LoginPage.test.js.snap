// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage snapshot 1`] = `
<div>
  <section
    class="layout-content"
  >
    <div
      class="loginPage"
    >
      <h1
        class="loginPage-title"
      >
        Log in to Nodepop
      </h1>
      <form
        class="loginForm"
      >
        <div
          class="formField loginForm-field"
        >
          <label
            class="formField-label"
          >
            <span>
              User Email
            </span>
            <input
              autocomplete="off"
              class="formField-input"
              name="email"
              type="text"
              value=""
            />
          </label>
        </div>
        <div
          class="formField loginForm-field"
        >
          <label
            class="formField-label"
          >
            <span>
              Password
            </span>
            <input
              autocomplete="off"
              class="formField-input"
              name="password"
              type="password"
              value=""
            />
          </label>
        </div>
        <input
          name="rememberCredentials"
          type="checkbox"
          value="rememberCredentials"
        />
        <label
          for="rememberCredentials"
        >
          Recordar credenciales
        </label>
        <div
          class="formButton"
        >
          <button
            class="submit-form-button btn btn-primary btn-lg"
            disabled=""
            type="submit"
          >
            Acceder
          </button>
        </div>
      </form>
    </div>
  </section>
</div>
`;
